spec.html

```diff
@@ -26277,6 +26308,91 @@ <h1>Object.prototype.valueOf ( )</h1>
         </emu-alg>
         <p>This function is the <dfn>%ObjProto_valueOf%</dfn> intrinsic object.</p>
       </emu-clause>
+
+      <emu-clause id="sec-object.prototype.__proto__" normative-optional>
+        <h1>Object.prototype.__proto__</h1>
+        <p>`Object.prototype.__proto__` is an accessor property with attributes { [[Enumerable]]: *false*, [[Configurable]]: *true* }. The [[Get]] and [[Set]] attributes are defined as follows:</p>
+
+        <emu-clause id="sec-get-object.prototype.__proto__">
+          <h1>get Object.prototype.__proto__</h1>
+          <p>The value of the [[Get]] attribute is a built-in function that requires no arguments. It performs the following steps when called:</p>
+          <emu-alg>
+            1. Let _O_ be ? ToObject(*this* value).
+            1. Return ? _O_.[[GetPrototypeOf]]().
+          </emu-alg>
+        </emu-clause>
+
+        <emu-clause id="sec-set-object.prototype.__proto__">
+          <h1>set Object.prototype.__proto__</h1>
+          <p>The value of the [[Set]] attribute is a built-in function that takes an argument _proto_. It performs the following steps when called:</p>
+          <emu-alg>
+            1. Let _O_ be ? RequireObjectCoercible(*this* value).
+            1. If Type(_proto_) is neither Object nor Null, return *undefined*.
+            1. If Type(_O_) is not Object, return *undefined*.
+            1. Let _status_ be ? _O_.[[SetPrototypeOf]](_proto_).
+            1. If _status_ is *false*, throw a *TypeError* exception.
+            1. Return *undefined*.
+          </emu-alg>
+        </emu-clause>
+      </emu-clause>
+
+      <emu-clause id="sec-object.prototype.__defineGetter__">
```

Comment:
Why would we want the legacy define methods to be normative required??

The vast majority of annex b should remain normative optional, and anything that’s definitely only needed on the web imo should stay in annex b.

Reply:
:shrug: no one had any reason to make these ones optional, unlike the `__proto__` accessor. even quickjs and xs implement `__defineGetter__` whatnot.

Reply:
#1JS

Reply:
The reason - one I hold for many parts of Annex B - is that these things are bad and should never have existed, and more importantly, are not universally used.

`__proto__` is used everywhere, in non-web code as well. `__defineGetter_` and friends are not.

Reply:
`__defineGetter__` and friends are sufficiently widely used that any implementation intending to run JavaScript as it exists in the world needs to implement them. Writing down those things is the whole point of having a standard. I don't see the point of omitting them.

Incidentally, that's also [what I understood to be consensus](https://github.com/tc39/ecma262/issues/1595) when this was discussed in plenary.

Reply:
I was [very explicit in plenary](https://github.com/tc39/notes/blob/7b2de881081abd34b02bc87bcdb662fd97555795/meetings/2019-10/october-1.md#conclusionresolution-4) that I was not providing consensus for carte-blanche moving everything out of Annex B. That link also includes in the checklist a clarification I gave to dan offline.

Reply:
I'm not sure what part of the notes you are intending to refer to by that link. All I see is that you made a request that we

> clearly mark things that we want to deprecate or discourage

which is fine by me, but is totally orthogonal to whether something is normative required or not.

Reply:
For (some) non-grammar things in Annex B, I'm saying I don't want them moved out of Annex B unless they are both normative optional and also explicitly "deprecated" in some form, since that's one of the current community effects of Annex B.

Reply:
I do not understand who we would be serving by making it normative optional.

Reply:
By making it required, we'd be signaling that it's a fine idea to use these things in new code. By marking it both deprecated and optional, a strong signal is sent that this isn't something anyone should be using in new code (which is basically the same signal I claim Annex B already sends).

Reply:
I don't see why "implementations are not required to have this (though they will find that they need to anyway if they want to run existing JavaScript), and you are discouraged from using this in new code" is significantly better on that score than "you are discouraging from using this in new code".

Also, I fairly strongly object to the idea that we should be making decisions about whether something is required or optional based on the signals we want to send to developers about the kinds of code we'd prefer them to write, rather than what implementations will need to implement to run JavaScript.

Reply:
How is `__proto__` different from, say, `with`, or `Symbol.species`?

Reply:
@mathiasbynens i'm not sure what you're asking - `__proto__` wasn't designed by the language's creator or standards committee, it was retroactively added because some browsers decided to add it themselves. `with` and `Symbol.species` were both intentionally designed and added. Separately, `__proto__` and `with` are grammar features (not counting the proto accessor), and I'm only discussing the API features.

@bakkot so to clarify, you are not on board with making these optional, but you're ok with marking them as deprecated/discouraged?

(Also, I'm not convinced that implementations actually *do* need to implement these particular things in order to run existing JavaScript; i'd be very surprised if most npm modules broke with these deleted from Object.prototype)

Reply:
@ljharb 

> so to clarify, you are not on board with making these optional, but you're ok with marking them as deprecated/discouraged

Yup.

>  i'd be very surprised if most npm modules broke with these deleted from Object.prototype

"most things wont break" is not the traditional standard. Many, many would. Node itself was using `__defineGetter__` internally until like v7. It's a part of JS, like it or not.

Reply:
I stand for optional normative. If the engine chooses to not adding it (Deno not implemented getter/setter `__proto__`), it's their own business. If the developer having problem with the missing `__proto__` or __defineGetter__`, it's still possible to define it in the userland.

Reply:
@ljharb I know the [history](https://blog.whatwg.org/javascript) of `__proto__`. The point I was making is that there's parts of the JavaScript standard that we now, in hindsight, consider to be mistakes (even outside of Annex B). Whether they were intentionally designed or organically became part of the standard after first becoming a de facto standard doesn't matter. It feels weird to special-case features just because they used to be in Annex B.

Reply:
FWIW, I wish `__proto__` was completely removed from specs, and from any engine. To date it caused more troubles than it supposed to solve, and both `Object` and `Reflect` have dedicated methods that replace it already.

The only usage I kinda think is reasonable, is for one-off literals creation, and literally nothing else: `{__proto__: Shared, counter: 0}` ... that's it.

Reply:
@ljharb so I just need to add a note that says we don't like them?

Reply:
The `__proto__` accessor and other **%Object.prototype%** `__`‑prefixed and suffixed properties should remain in **Annex B**, as they have the much saner `{Object,​Reflect}.{defineProperty,​getOwnPropertyDescriptor,​getPrototypeOf,​setPrototypeOf}` alternatives.

Reply:
@mathiasbynens that's totally true, but the entire purpose of Annex B was always "this is a special case we don't want in the actual language, and we don't want people using, but is here for web compat". In other words, I'm saying that "being in Annex B" is *already* being special-cased, and i don't want that indiscriminately lost.

@devsnek just like "normative optional" is a special tag, i'd want some kind of "deprecated" tag, that referenced a big scary note saying these are bad and if you use them you should feel bad, yes.